overlord/configstate/config/transaction.go

```diff
@@ -343,6 +381,130 @@ func commitChange(pristine *json.RawMessage, change interface{}) *json.RawMessag
+		}
+	}
+	return true, nil
+}
+
+// mergeConfigWithVirtual takes the given configuration and merges it
+// with the virtual configuration values by calling the registered
+// virtual configuration function of the given snap. The merged config
+// is returned.
+func mergeConfigWithVirtual(instanceName, requestedKey string, origConfig *map[string]*json.RawMessage) error {
+	virtualMu.Lock()
+	km, ok := virtualMap[instanceName]
+	virtualMu.Unlock()
+	if !ok {
+		return nil
+	}
+
+	// create a "patch" from the virtual entries
+	patch := make(map[string]interface{})
+	for virtualKey, virtualFn := range km {
+		if virtualFn == nil {
+			continue
+		}
+		// check if the requested key is part of the virtual
+		// configuration
+		partOf, err := overlapsWithVirtualConfig(requestedKey, virtualKey)
+		if err != nil {
+			return err
+		}
+		if !partOf {
+			continue
+		}
+
+		// Pass the right key to the virtualFn(), this can
+		// either be a subtree of the virtual-tree or the
+		// other virtualKey itself.
+		k := requestedKey
+		if len(requestedKey) < len(virtualKey) {
+			k = virtualKey
+		}
+		res, err := virtualFn(k)
+		if err != nil {
+			return err
+		}
+		patch[virtualKey] = jsonRaw(res)
+	}
+	if len(patch) == 0 {
+		return nil
+	}
+
+	// create a "working copy" of the config and apply the patches on top
+	config := jsonRaw(*origConfig)
+	patchKeys := sortPatchKeysByDepth(patch)
+	for _, subkeys := range patchKeys {
+		// patch[key] above got assigned jsonRaw() so this cast is ok
+		raw := patch[subkeys].(*json.RawMessage)
+		mergedConfig, err := PatchConfig(instanceName, strings.Split(subkeys, "."), 0, config, raw)
+		if err != nil {
+			return err
+		}
+		// PatchConfig got *json.RawMessage as input and
+		// returns the same type so this cast is ok (but be defensive)
+		config, ok = mergedConfig.(*json.RawMessage)
+		if !ok {
+			return fmt.Errorf("internal error: PatchConfig in mergeConfigWithVirtual did not return a *json.RawMessage please report this as a bug")
+		}
+	}
+
+	// XXX: unmarshaling on top of something leaves values in place
+	// (no problem here because we only add virtual things)
+	// convert back to the original config
+	if err := jsonutil.DecodeWithNumber(bytes.NewReader(*config), origConfig); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// clearVirtualConfig iterates over a given config and removes any values
+// that come from virtual configuration. This is used before committing a
+// config to disk.
+func clearVirtualConfig(instanceName string, snapChanges map[string]interface{}) {
+	virtualMu.Lock()
+	km := virtualMap[instanceName]
+	virtualMu.Unlock()
+
+	clearVirtualConfigRecursive(km, snapChanges, "")
+}
+
+func clearVirtualConfigRecursive(km map[string]VirtualCfgFunc, config map[string]interface{}, keyprefix string) {
+	if len(keyprefix) > 0 {
+		keyprefix += "."
+	}
+	for key, value := range config {
+		// any top-level virtual keys are removed
+		if _, ok := km[keyprefix+key]; ok {
+			delete(config, key)
```

Comment:
can't we just `continue` here, as we have removed the entire key, it doesn't matter to clean value?

Reply:
Indeed we can, nice catch!